frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+			Identifier = <T as Config>::Identifier,
+		>,
+	>,
+>;
+
+/// Convenience wrapper for [`MigrationCursor`].
+pub type CursorOf<T> =
+	MigrationCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+/// Convenience wrapper for [`ActiveCursor`].
+pub type ActiveCursorOf<T> =
+	ActiveCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// All the multi-block migrations to run.
+		///
+		/// Should only be updated in a runtime-upgrade once all the old ones have completed. (Check
+		/// `Cursor` for `None`).
+		type Migrations: Get<MigrationsOf<Self>>;
+
+		/// The cursor type that is shared across all migrations.
+		type Cursor: FullCodec + MaxEncodedLen + TypeInfo + Parameter;
+
+		/// The identifier type that is shared across all migrations.
+		type Identifier: FullCodec + MaxEncodedLen + TypeInfo;
+
+		/// Notification handler for status updates regarding runtime upgrades.
+		///
+		/// Can be used to pause XCM etc.
+		type UpgradeStatusHandler: UpgradeStatusHandler;
+
+		/// The weight to spend each block to execute migrations.
+		type ServiceWeight: Get<Weight>;
+
+		/// Weight information for the calls and functions of this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The currently active migration to run and its cursor.
+	///
+	/// `None` indicates that no migration process is running.
+	#[pallet::storage]
+	pub type Cursor<T: Config> = StorageValue<_, CursorOf<T>, OptionQuery>;
+
+	/// Set of all successfully executed migrations.
+	///
+	/// This is used as blacklist, to not re-execute migrations that have not been removed from the
+	/// codebase yet. Governance can regularly clear this out via `clear_historic`.
+	#[pallet::storage]
+	pub type Historic<T: Config> = StorageMap<_, Twox64Concat, T::Identifier, (), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Runtime upgrade started.
+		UpgradeStarted,
+		/// Runtime upgrade completed with `migrations`.
+		UpgradeCompleted { migrations: u32 },
+		/// Runtime upgrade failed.
+		///
+		/// This is very bad and will require governance intervention.
+		UpgradeFailed,
+		/// Migration `index` was skipped, since it already executed in the past.
+		MigrationSkippedHistoric { index: u32 },
+		/// Migration `index` made progress.
+		MigrationAdvanced { index: u32, step: T::BlockNumber },
+		/// Migration `index` completed.
+		MigrationCompleted { index: u32, took: T::BlockNumber },
+		/// Migration `index` failed.
+		///
+		/// This implies that the whole upgrade failed and governance intervention is required.
```

Comment:
Maybe we could add a convenience call to make recovery easier that resets storage & the cursor? called something like `reset`, callable by sudo.

Reply:
How would we reset storage? Only the cursor needs to be reset to unstuck it.  
I also dont think that we need to optimize for this case - since it should not be needed anyway.  

This makes me think if we could add a `undo` function to the migrations, or even create a storage layer over multiple blocks, that only commits at the end on success...

Reply:
True, maybe not worth the convenience method if it's just resetting the cursor. 

Storage layer over multiple blocks would be very cool - perhaps a future improvement?

Reply:
It also sounds incredibly non-trivial, and I think there's a lot more important similarly non-trivial stuff to be getting on with in the meantime.